source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,910 @@
+
+  ir::Instruction* new_constant = nullptr;
+
+  // If the remainder is negative then we add a signed constant, otherwise just
+  // add an unsigned constant.
+  if (remainder < 0) {
+    new_constant =
+        builder.Add32BitSignedIntegerConstant(static_cast<int32_t>(remainder));
+  } else {
+    new_constant = builder.Add32BitUnsignedIntegerConstant(
+        static_cast<int32_t>(remainder));
+  }
+
+  uint32_t constant_id = new_constant->result_id();
+
+  // Add the merge block to the back of the binary.
+  blocks_to_add_.push_back(
+      std::unique_ptr<ir::BasicBlock>(new_loop->GetMergeBlock()));
+
+  AddBlocksToLoop(new_loop);
+  // Add the blocks to the function.
+  AddBlocksToFunction(loop->GetMergeBlock());
+
+  // Reset the usedef analysis.
+  context_->InvalidateAnalysesExceptFor(
+      ir::IRContext::Analysis::kAnalysisLoopAnalysis);
+  opt::analysis::DefUseManager* def_use_manager = context_->get_def_use_mgr();
+
+  // Update the condition check.
+  ir::Instruction* condition_check = def_use_manager->GetDef(
+      original_conditional_branch.GetSingleWordOperand(0));
+
+  // This should have been checked by the LoopUtils::CanPerformUnroll function
+  // before entering this.
+  assert(condition_check->opcode() == SpvOpSLessThan);
+  condition_check->SetInOperand(1, {constant_id});
+
+  // Update the next phi node. The phi will have a constant value coming in from
+  // the preheader block. For the duplicated loop we need to update the constant
+  // to be the amount of iterations covered by the first loop and the incoming
+  // block to be the first loops new merge block.
+  uint32_t phi_incoming_index =
+      GetPhiIndexFromLabel(loop->GetPreHeaderBlock(), loop_induction_variable_);
+  loop_induction_variable_->SetInOperand(phi_incoming_index - 1, {constant_id});
+  loop_induction_variable_->SetInOperand(phi_incoming_index, {new_merge_id});
+
+  context_->InvalidateAnalysesExceptFor(
+      ir::IRContext::Analysis::kAnalysisLoopAnalysis);
+
+  context_->ReplaceAllUsesWith(loop->GetMergeBlock()->id(), new_merge_id);
+
+  ir::LoopDescriptor& loop_descriptor =
+      *context_->GetLoopDescriptor(&function_);
+
+  loop_descriptor.AddLoop(new_loop, loop->GetParent());
+}
+
+// Duplicate the |loop| body |factor| number of times while keeping the loop
+// backedge intact.
+void LoopUnrollerUtilsImpl::PartiallyUnroll(ir::Loop* loop, size_t factor) {
+  Unroll(loop, factor);
+  AddBlocksToLoop(loop);
+  AddBlocksToFunction(loop->GetMergeBlock());
+}
+
+// Duplicate the |loop| body |factor| number of times while keeping the loop
+// backedge intact.
+void LoopUnrollerUtilsImpl::Unroll(ir::Loop* loop, size_t factor) {
+  state_ = LoopUnrollState{loop_induction_variable_, loop->GetLatchBlock(),
+                           loop_condition_block_};
+  for (size_t i = 0; i < factor - 1; ++i) {
+    CopyBody(loop, true);
+  }
+
+  uint32_t phi_index = GetPhiIndexFromLabel(state_.previous_continue_block_,
```

Comment:
If I got this right, this section is going back to the top to rewrite the OpPhi's value that comes from the backedge.  The issue that @dnovillo brought up earlier (multiple induction variables) is that there may be multiple OpPhi instructions that need to be rewritten.  Where are the other OpPhi instructions rewritten?

Reply:
They won't be rewritten in this patch as it is, we check in the CanPerformUnroll function that the only phi used in the loop is the induction variable. We have a solution to rewrite the phis after we transform the loop into LCSAA and give it a single preheader. After doing that we can fairly easily extend the induction variable rewriting part to rewrite all the phis found. 

We could add that to this patch if required, otherwise it will be a followup PR. I added a note to the CanPerformUnroll method in LoopUtils.h to list the constraints we currently enforce.